### ISSUE TITLE
NoReturn return type from __init__ ignored

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#7757]
**Describe the bug**
A class with an `__init__` method with a return type of `NoReturn` is ignored at the call-site. This is similar to #3188, but Mypy now supports this pattern as of https://github.com/python/mypy/issues/10342, so I think it's reasonable to revisit this pattern.

**Code or Screenshots**
[Pyright playground](https://pyright-play.net/?pythonVersion=3.11&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByYASgKYwCuIKAsAFBY55QDOcL9ndAJqcFMDBgAFAEooAWgB8hEuSooAXPSirW7AHSkAHkhjCAjKK70AxgBsAhixZQAgsrpqovfgH03qPR%2BEtS54HFpWTJKakdnZ0ERYzouVygAI0sQYW0gmRYYEAi1JH5tXMioEHlqKAAiCpU1AGIBITFVeoBhXFLTGHM4VhgwBFsABTgQJDQAC3wWSzhUDBhx0hKylChTMApzbiSlohRSGtU7Jqh6gElbIlCFJIp8FDB8XhhSTtJtxdKknuHRiZh6EA) (doesn't understand `NoReturn` from `__init__`)

[Mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=71955c8c44161e8a805e80f02ceb355a) (works correctly)

```python
from typing import NoReturn
import sys


def foo() -> NoReturn:
    sys.exit(1)


class A:
    def __init__(self) -> NoReturn:
        foo()


def bar(x) -> str:
    if x:
        return ""
    # foo()  # Correctly stops Pyright saying the return could be None
    A()  # Is NoReturn but not detected here by Pyright
```

**VS Code extension or command-line**
Found using commandline version. Recreated in playground in most recent version.

**Comment by erictraut** [comment]
The [Python typing spec](https://typing.readthedocs.io/en/latest/spec/constructors.html) was recently updated to clarify how constructor calls are meant to be evaluated by a type checker. I was the primary author of this new chapter, but it went through [significant review](https://discuss.python.org/t/draft-typing-spec-chapter-for-constructors/49744/26) and [approval by the full typing council](https://github.com/python/typing-council/issues/25). 

Pyright is conformant with the agreed-upon spec, but mypy [diverges from it pretty significantly](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) in the area of constructors.

The current spec does not recognize that `__init__` can return anything other than `None`. I think there's a reasonable argument to be made in favor of supporting this, but the typing spec would need to be amended (again) to allow this. If the spec is changed, I will update pyright accordingly. If the spec is not changed, I'm not going to update pyright because it would then deviate from the spec.

If you would like to make the case for this change to the spec, please start a new discussion in the [Python typing forum](https://discuss.python.org/c/typing/32). To convince the community (and the typing council) that such a change should be made, you'll need to make a case for why such a change is needed.

For now, I'm going to reject this change and close the issue, but I'm willing to re-open it if and when the typing spec is modified.
